### DISCUSSION TITLE
Websocket: Way to easily remove/replace it in the code?

### DISCUSSION THREAD

**Discussion by retohofstetter** [lllyasviel/Fooocus#1892] [General]
Hey,

I figured Foocus is using websocket when displaying the images. Is there a way around that? Our firewall is not letting websocket unfortunately...

Best, Reto

**Comment by aolko** [comment]
use gradio 4? They moved from websockets to SSE iirc

**Reply by retohofstetter** [reply]
I upgraded to gradio v4 in the embedded python, but then it crashes. Wonder what else would need to be changed to make it work with Gradio v4?

Traceback (most recent call last):
  File "D:\fooocus\Fooocus\launch.py", line 112, in <module>
    from webui import *
  File "D:\fooocus\Fooocus\webui.py", line 13, in <module>
    import modules.gradio_hijack as grh
  File "D:\fooocus\Fooocus\modules\gradio_hijack.py", line 21, in <module>
    from gradio.components.base import IOComponent, _Keywords, Block
ImportError: cannot import name 'IOComponent' from 'gradio.components.base' (D:\fooocus\python_embeded\lib\site-packages\gradio\components\base.py)

**Reply by mashb1t** [reply]
There is an old, outdated PR, which may help guide you: https://github.com/lllyasviel/Fooocus/pull/961/files
Please be aware of that this could take a few days to upgrade and make everything compatible. Feel free to open another PR when you're done 👍

**Reply by retohofstetter** [reply]
Cool, will try to figure it out!

**Comment by retohofstetter** [comment]
Hm, Foocus does not use gradio 4, do they? At least the current version uses websockets still, no? But could try to use gradio 4 and see what happens with Foocus :) (hope not crashing), thanks for the comment!

**Reply by mashb1t** [reply]
Fooocus uses Gradio 3.41.2 with websockets.
You might have a look at https://github.com/konieshadow/Fooocus-API, which uses a REST API, which could be the solution for you.

**Reply by retohofstetter** [reply]
The API is awesome! But that would require rewriting the whole frontend, which is great with gradio as is...
